[Constraint System]  Fix covariant erasure for constrained existentials

angela-laar opened the pull request:
Constrained existentials should be type erased to an upper bound that is dependent on other type parameters.

```swift
let t: (any Sequence<Int>) = [1, 2]

for value in t {
    let v = value + 1 // the existential is opened so that `value` is generic parameter Int instead of Any
}
```

Fixes rdar://101343646

angela-laar committed:
[Constraint System]  Fix covariant erasure for constrained existentials

Use of an associated type in the member is not considered invariant
if that associated type has been made concrete by the existential type.

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -1568,8 +1568,7 @@ shouldOpenExistentialCallArgument(
       /*skipParamIdx=*/paramIdx);
-  if (referenceInfo.selfRef > TypePosition::Covariant ||
-      referenceInfo.assocTypeRef > TypePosition::Covariant)
+  if (referenceInfo.selfRef > TypePosition::Covariant)
```
What does this change do?

angela-laar replied:
This is to fix opening constrained existential with opaque result type: 

```
 func foo<T>(_ c: some Collection<T>) -> some Collection<T> {
  return c
}
let c: any Collection<Int> = [1, 2, 3] 
let result = foo(c) // causes Type 'any Collection<Int>' cannot conform to 'Collection' error
```

I thought this would fix my failing test but looks like this is a separate issue. Going to make a new PR  for this 🙂

slavapestov commented on test/type/parameterized_existential.swift:
```diff
@@ -91,3 +91,9 @@ func protocolCompositionNotSupported1(_: SomeProto & Sequence<Int>) {}
 func protocolCompositionNotSupported2(_: any SomeProto & Sequence<Int>) {}
 // expected-error@-1 {{non-protocol, non-class type 'Sequence<Int>' cannot be used within a protocol-constrained type}}
+
+func increment(n : any Sequence<Float>) {
```
Can you also test the case where `func` is generic and the argument to the existential contains a generic parameter type?

angela-laar commented:
@swift-ci please smoke test macOS

angela-laar committed:
[test] Test covariant erasure for constrained existentials

And update expectederror for member access test

angela-laar commented:
@swift-ci please smoke test

AnthonyLatsis commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2103,7 +2103,7 @@ typeEraseExistentialSelfReferences(
         erasedTy = baseTy;
       } else {
-        erasedTy = existentialSig->getNonDependentUpperBounds(t);
+        erasedTy = existentialSig->getDependentUpperBounds(t);
```
What if:

```swift
protocol P<A> {
  associatedtype A
}
protocol Q {
  associatedtype B
  associatedtype C: P<B>

  func foo() -> C
}

func test(q: any Q) {
  let _ = q.foo() // Currently erased to 'any P', but would be 'any P<B>'?
}
```

The erased type should not be dependent on the opened existential. Instead, I think we should keep using `getNonDependentUpperBounds`, but refactor it to type-erase the result of calling `getDependentUpperBounds`.

angela-laar replied:
Oh wait, with and without this change, that example will still produce this error: `Inferred result type 'any P' requires explicit coercion due to loss of generic requirements`

hborla replied:
> Either way we shouldn’t be assuming that getDependentUpperBounds will always produce a non-dependent result; it will break stuff one way or the other

I consider the fact that your example compiles today to be a bug. This code should require an explicit coercion to `as C1`, which is described in the ["Losing" constraints when type-erasing resulting values](https://github.com/apple/swift-evolution/blob/main/proposals/0352-implicit-open-existentials.md#losing-constraints-when-type-erasing-resulting-values) section of SE-0352. This rule also applies to SE-0309, as mentioned at the end of the section.

AnthonyLatsis replied:
> I consider the fact that your example compiles today to be a bug.

I agree (well spotted!). The problem is that this change causes it to crash down the line due to the unilateral contract breakage. In regards to Angela’s original example, I think a more principled approach would be to look into why the function is being passed `any IteratorProtocol` rather than `any IteratorProtocol<Int>` as the base type.

slavapestov replied:
Another thing I noticed is that typeEraseExistentialSelfReferences() is basically a copy-and-paste of typeEraseOpenedArchetypesWithRoot() from Type.cpp. Perhaps if we update both copies or merge them together it will fix this problem.

xedin replied:
> I think a more principled approach would be to look into why the function is being passed `any IteratorProtocol` rather than `any IteratorProtocol<Int>` as the base type.

@AnthonyLatsis  If I understand your comment correctly by "the function" you mean `typeEraseExistentialSelfReferences` and considering `next()` reference.

The problem is with `makeIterator` reference, that's why `next()` doesn't get `any IteratorProtocol<Int>` for the base. I simplified the example down to:

```swift
func increment(n : any Sequence<Float>) {
 var iter = n.makeIterator()
}
```
(the solver transforms `for-in` loop into `var $gen = <expr>.makeIterator(); while let v = $gen.next() { ... }` so `$gen` is `n` in this example).

`typeEraseExistentialSelfReferences` gets `any Sequence<Float>` as a base type and `refTy` is `() -> τ_0_0.Iterator`, generic signature is `<τ_0_0 where τ_0_0 : Sequence, τ_0_0.Element == Float>`.

Since the transform runs on `refTy` it would try to call `getConcreteType` on `τ_0_0.Iterator` and that does not produce a concrete type so it falls through to erasure case and produces `any IteratorProtocol`. 

> I think we should keep using `getNonDependentUpperBounds`, but refactor it to type-erase the result of calling `getDependentUpperBounds`.

Let me see if I understand your point here - `getNonDependentUpperBounds` should be allowed to use use type information provided by primary associated types, _but only them_, to maintain the contract that visible type retains only type information as available before the reference otherwise we might end up in the situation that leaks type information from opened type to the outside. In practice it means that `getNonDependentUpperBounds` needs to keep erasing until it reaches bound that does not have any unbound dependent members/generic parameters (based on the existential signature), right?

For your second example this means that `getNonDependentUpperBounds` would produce `C1` (because `C2` references undetermined dependent member `A`) but if `P` did have a primary associated type `A` and it was bound in expression i.e. `any P<Int>` `getNonDependentUpperBounds` would produce `C2<Int>` instead, does that make sense?

AnthonyLatsis replied:
Sorry, this got lost among CI commands in the mail thread.

> Let me see if I understand your point here - `getNonDependentUpperBounds` should be allowed to use use type information provided by primary associated types, _but only them_, to maintain the contract

`getNonDependentUpperBounds` is allowed to use all the type info provided by the receiver (the generic signature derived from the existential base type). All generic constraints contribute to the result, e.g.

```swift
class C<T> {}
protocol P { 
  associatedtype A 
  associatedtype B: C<A>
}
protocol Q: P where A == Int {
  func getB() -> B
}

do {
  let q: any Q
  let _ = q.getB() // OK, 'C<Int>'
}
```
`getNonDependentUpperBounds` produces a constraint type that is *supposed* to be the non-dependent supremum of the dependent upper bounds, which retain all the type info. In practice and by contract, it’s always a non-dependent type, but not always as specific as it could have been. For example, the method wasn’t updated for constrained existentials, so it doesn’t know how to build parameterized protocol types. This means it never produces a type containing a parameterized protocol type; just bare protocols.

> In practice it means that `getNonDependentUpperBounds` needs to keep erasing until it reaches bound that does not have any unbound dependent members/generic parameters (based on the existential signature), right?

> For your second example this means that `getNonDependentUpperBounds` would produce `C1` (because `C2` references undetermined dependent member `A`) but if `P` did have a primary associated type `A` and it was bound in expression i.e. `any P<Int>` `getNonDependentUpperBounds` would produce `C2<Int>` instead, does that make sense?

Yes.

___

> `typeEraseExistentialSelfReferences` gets `any Sequence<Float>` as a base type and `refTy` is `() -> τ_0_0.Iterator`, generic signature is `<τ_0_0 where τ_0_0 : Sequence, τ_0_0.Element == Float>`.
> 
> Since the transform runs on `refTy` it would try to call `getConcreteType` on `τ_0_0.Iterator` and that does not produce a concrete type so it falls through to erasure case and produces `any IteratorProtocol`. 

I see, that makes a lot of sense. So after all this suggests we have to teach `getNonDependentUpperBounds` to build parameterized protocol types (`getDependentUpperBounds` already does this), and probably do so on demand via a flag because — I presume — upgrading upper bounds for more than just the built-in `for-in` transform would compromise the `requires explicit coercion` diagnostic.

angela-laar replied:
ok! will do

angela-laar replied:
> I consider the fact that your example compiles today to be a bug.

I will track fixing this diagnostic issue in a second PR since after a couple of attempts, it seems significant changes to the Coercion Checker are needed

Opened a new PR for the original issue: https://github.com/apple/swift/pull/65785

xedin reviewed (approved):
Please add more tests before merging as we've discussed.

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test macOS

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[Sema] Erase existential PATs to reduced type

When opening an existential return or parameter type that has a
primaray associated type, the primmary asssociated type should be
preserved if the type can be reduced to a concrete type. Otherwise,
we should erase the existential to a type with no dependent members
or generic parameters and diagnose explicit coercion is needed.

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[test] fix tests

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
> I consider the fact that your example compiles today to be a bug.


I will track fixing this diagnostic issue in a second PR since after a couple of attempts, it seems significant changes to the Coercion Checker are needed

Opened a new PR for the original issue: https://github.com/apple/swift/pull/65785
